### PR TITLE
#3212 Pack variants in Returns

### LINKED ISSUE #3212: Add pack variant components for returns (3h)
## Is your feature request related to a problem? Please describe 👀

Pack variants has been merged! The code for pack variants is in our feature branch, but not yet added to the returns UI pieces.

If I'm returning some goods, or receiving a return, I want to be able to specify the pack variant of the goods.

## Describe the solution you'd like 🎁

Replace the PackSize column with the new `PackVariantCell` - any UI tweaks required?

### PR DESCRIPTION
**Pull request by CarlosNZ**
Fixes #3212 

Please note: incorrect issue number in branch name -- oops 🙄

# 👩🏻‍💻 What does this PR do? 
 

I'm not sure that this is working exactly how it "should" (I don't fully understand it), but it seems to be the same as the equivalent in Outbound/Inbound shipments as far as I can tell. So any further refinements should probably be done to both shipments and returns at the same time (after further clarification?) -- do we need a new issue?

# 🧪 How has/should this change been tested? 

Have some items with pack variants.

For Outbound returns:
- DetailView table should show "Pack", "Pack Qty" and "UnitQty" columns
- Should display the "defaultPackSize" for that invoice line
- Modal should display the same (this is unchanged)
- Confirm appearance and functionality the same as Outbound Shipments (for modal and table)

For Inbound returns:
- DetailView table should show "Pack", "Pack Qty" and "UnitQty" columns
- Should display in the base unit
- Modal should display in base unit, but can be switched for display purposes (unchanged in this PR)
- Confirm appearance and functionality the same as Inbound Shipments (for modal and table)

## 💌 Any notes for the reviewer?

Let me know if I've misunderstood the requirements here -- not sure how much should be done in this PR, hence I've just got it to parity with Shipments.

### CODE REVIEW

File: client/packages/common/src/utils/arrays/ArrayUtils.ts [hunk 1]
```diff
@@ -38,7 +38,7 @@ export const ArrayUtils = {
     return arr.reduce((sum, someEntity) => sum + someEntity[key], 0);
   },
   // De-duplicate (remove duplicates)
-  dedup: uniq,
+  dedupe: uniq,
```
Comment (CarlosNZ) [inline client/packages/common/src/utils/arrays/ArrayUtils.ts:41]: Unrelated typo fix

File: client/packages/invoices/src/Returns/api/operations.generated.ts [hunk 2]
```diff
@@ -262,6 +262,14 @@ export const OutboundReturnLineFragmentDoc = gql`
...
   packSize
   sellPricePerPack
+  item {
+    __typename
+    id
+    name
+    code
+    unitName
+    defaultPackSize
+  }
```
Comment (CarlosNZ) [inline client/packages/invoices/src/Returns/api/operations.generated.ts:272]: Copied from the equivalent in shipments.

File: client/packages/system/src/Item/Components/PackVariant/PackVariantEntryCell.tsx [hunk 3]
```diff
@@ -15,7 +15,9 @@ import { usePackVariant } from '../../context';
 const ENTER_PACK_SIZE = -1;
 export const PACK_VARIANT_ENTRY_CELL_MIN_WIDTH =
   150 + DEFAULT_NUMERIC_TEXT_INPUT_WIDTH;
-// This cell displays a packSize number input and unit pack drop down if unit pack variants exist
+// This cell displays a packSize number input and unit pack drop down if unit
```
Comment (CarlosNZ) [inline client/packages/system/src/Item/Components/PackVariant/PackVariantEntryCell.tsx:18]: Auto-formatting only in this file

File: client/packages/invoices/src/Returns/InboundDetailView/columns.ts [hunk 4]
```diff
@@ -151,6 +152,26 @@ export const useInboundReturnColumns = ({
...
+          getItemId: row => {
+            if ('lines' in row) return '';
+            else return row?.item?.id;
+          },
+          getPackSizes: row => {
+            if ('lines' in row) return row.lines.map(l => l.packSize ?? 1);
+            else return [row?.packSize ?? 1];
+          },
+          getUnitName: row => {
+            if ('lines' in row) return null;
```
Comment (lache-melvin) [inline client/packages/invoices/src/Returns/InboundDetailView/columns.ts:169]:
```suggestion
            if ('lines' in row) return row.lines[0]?.item?.unitName ?? null;
```
Any reason this isn't the same as Outbound Return? When the IR is not grouped by item:
Screenshot 2024-04-05 at 8 48 50 AM

When grouped by item:
Screenshot 2024-04-05 at 8 48 42 AM

When not showing [multiple] I think it should still show the unit?

**Review by lache-melvin — CHANGES_REQUESTED** [review]
Works great for me! Thanks Carl, just the one grouped by item issue to fix (or if there's a reason you left as null?)

File: client/packages/invoices/src/Returns/InboundDetailView/columns.ts (hunk 4 above)
Comment (CarlosNZ) [reply client/packages/invoices/src/Returns/InboundDetailView/columns.ts:169]: Yeah, except this is how it is in Inbound Shipment -- I just copied that. Should we leave it for now and change them both once we know exactly what the requirement is?

Comment (lache-melvin) [reply client/packages/invoices/src/Returns/InboundDetailView/columns.ts:169]: Just checked with Andrei - says both should show the unit name 🤓

Comment (CarlosNZ) [reply client/packages/invoices/src/Returns/InboundDetailView/columns.ts]: Righto, made that change to both. Thanks a81185b38

**Review by lache-melvin — APPROVED** [review]
Looks great Carl, thanks for changes 🚀
